File: writing-for-humans-workspace/eval.ts
import { Database } from "bun:sqlite";
import { createHash } from "node:crypto";
import {
  mkdir,
  mkdtemp,
  readFile,
  writeFile,
  readdir,
  realpath,
} from "node:fs/promises";
import { homedir } from "node:os";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const workspace = dirname(fileURLToPath(import.meta.url));
const root = dirname(workspace);
const [command, name] = process.argv.slice(2);
if (!name || !/^iteration-[1-9]\d*$/.test(name))
  throw new Error(
    "Usage: bun writing-for-humans-workspace/eval.ts <prepare|start|status|capture> iteration-N",
  );
const dir = join(workspace, "live", name);
const readJSON = async (path: string) =>
  JSON.parse(await readFile(path, "utf8"));
const saveJSON = (path: string, value: unknown) =>
  writeFile(path, JSON.stringify(value, null, 2) + "\n");
const hash = (text: string) => createHash("sha256").update(text).digest("hex");
const db = () =>
  new Database(
    process.env.WRITING_EVAL_DB || join(homedir(), ".t3/userdata/state.sqlite"),
    {
      readonly: true,
    },
  );
function git(...args: string[]) {
  const result = Bun.spawnSync(["git", ...args], { cwd: root });
  if (result.exitCode) throw new Error(result.stderr.toString());
  return result.stdout.toString().trim();
}
async function verifyInputs(run: any) {
  for (const [file, digest] of Object.entries(run.hashes))
    if (hash(await readFile(join(dir, file), "utf8")) !== digest)
      throw new Error(`Saved input changed: ${file}`);
}
function inspect(threadId: string) {
  const connection = db();
  try {
    return {
      thread: connection
        .query(
          "SELECT thread_id, title, worktree_path, model_selection_json, runtime_mode, interaction_mode, deleted_at FROM projection_threads WHERE thread_id = ?",
        )
        .get(threadId) as any,
      session: connection
        .query(
          "SELECT status, active_turn_id, last_error, provider_name FROM projection_thread_sessions WHERE thread_id = ?",
        )
        .get(threadId) as any,
      turns: connection
        .query(
          "SELECT turn_id, state, started_at, completed_at FROM projection_turns WHERE thread_id = ? ORDER BY requested_at, row_id",
        )
        .all(threadId),
      messages: connection
        .query(
          "SELECT message_id, turn_id, role, text, is_streaming, created_at, updated_at, attachments_json FROM projection_thread_messages WHERE thread_id = ? ORDER BY created_at, rowid",
        )
        .all(threadId) as any[],
    };
  } finally {
    connection.close();
  }
}

if (command === "prepare") {
  const files: Record<string, string> = {};
  for (const file of ["topic.md", "settings.json"])
    files[file] = await readFile(join(workspace, file), "utf8");
  for (const skill of ["writing-for-humans", "unslop"])
    files[`${skill}.md`] = await readFile(
      join(root, "skills/software-development", skill, "SKILL.md"),
      "utf8",
    );
  files["prompt.txt"] =
    `${files["topic.md"].trim()}\n\nUse $writing-for-humans and $unslop.\n`;
  const settings = JSON.parse(files["settings.json"]);
  if (!settings.model || !settings.effort || !settings.provider)
    throw new Error("Incomplete participant settings");
  await mkdir(join(workspace, "live"), { recursive: true });
  await mkdir(dir); // Never overwrite a run.
  // A separate repo has no authoring-repo ancestry, instructions, or remote.
  const participantDirectory = await realpath(await mkdtemp("/tmp/team-app-"));
  const init = Bun.spawnSync([
    "git",
    "init",
    "-q",
    "-b",
    "main",
    participantDirectory,
  ]);
  if (init.exitCode) throw new Error(init.stderr.toString());
  for (const skill of ["writing-for-humans", "unslop"]) {
    const target = join(participantDirectory, ".agents/skills", skill);
    await mkdir(target, { recursive: true });
    await writeFile(join(target, "SKILL.md"), files[`${skill}.md`]);
  }
  for (const [file, text] of Object.entries(files))
    await writeFile(join(dir, file), text);
  await saveJSON(join(dir, "run.json"), {
    createdAt: new Date().toISOString(),
    name: "Team invitations",
    status: "prepared",
    sourceRevision: git("rev-parse", "HEAD"),
    sourceBranch: git("branch", "--show-current"),
    branch: "main",
    directory: participantDirectory,
    context:
      "Standalone temporary repo with only the two skill files. No evaluation material or authoring-repo instructions. Normal harness and filesystem permissions still apply.",
    settings,
    hashes: Object.fromEntries(
      Object.entries(files).map(([file, text]) => [file, hash(text)]),
    ),
    threadId: null,
  });
  console.log(`Prepared ${dir}`);
} else {
  const run = await readJSON(join(dir, "run.json"));
  await verifyInputs(run);
  if (command === "start") {
    if (run.status !== "prepared" || run.threadId)
      throw new Error(
        "Launch already attempted. Use status and inspect dispatch.json before taking any recovery action.",
      );
    for (const skill of ["writing-for-humans", "unslop"]) {
      const text = await readFile(
        join(run.directory, ".agents/skills", skill, "SKILL.md"),
        "utf8",
      );
      if (hash(text) !== run.hashes[`${skill}.md`])
        throw new Error(`Participant skill changed before launch: ${skill}`);
    }
    const s = run.settings;
    const args = [
      "python3",
      join(root, "skills/system/to-thread/scripts/t3-thread.py"),
      "--name",
      run.name,
      "--prompt",
      await readFile(join(dir, "prompt.txt"), "utf8"),
      "--project-directory",
      root,
      "--directory",
      run.directory,
      "--branch",
      run.branch,
      "--provider",
      s.provider,
      "--model",
      s.model,
      "--effort",
      s.effort,
      "--runtime-mode",
      s.runtimeMode,
      "--interaction-mode",
      s.interactionMode,
    ];
    if (s.serviceTier) args.push("--service-tier", s.serviceTier);
    run.status = "launch-attempted";
    await saveJSON(join(dir, "run.json"), run);
    const child = Bun.spawn(args, {
      cwd: root,
      stdout: "pipe",
      stderr: "pipe",
    });
    const [stdout, stderr, code] = await Promise.all([
      new Response(child.stdout).text(),
      new Response(child.stderr).text(),
      child.exited,
    ]);
    await writeFile(join(dir, "dispatch.stdout.txt"), stdout);
    await writeFile(join(dir, "dispatch.stderr.txt"), stderr);
    if (stdout.trim()) {
      const result = JSON.parse(stdout);
      await saveJSON(join(dir, "dispatch.json"), result);
      run.threadId = result.thread_id ?? null;
    }
    run.status = code ? "launch-needs-inspection" : "launched-unverified";
    await saveJSON(join(dir, "run.json"), run);
    if (code)
      throw new Error(
        `Launch needs inspection: ${stderr}. Preserve any thread identity in run.json.`,
      );
    console.log(
      `Started ${run.name}: ${run.threadId}. Use status to verify it.`,
    );
  } else if (command === "status") {
    if (!run.threadId)
      throw new Error("No thread identity recorded; inspect dispatch output");
    const state = inspect(run.threadId);
    const selection = state.thread
      ? JSON.parse(state.thread.model_selection_json)
      : null;
    const matches =
      selection?.instanceId === run.settings.provider &&
      selection?.model === run.settings.model &&
      selection?.options?.some(
        (o: any) =>
          o.id === "reasoningEffort" && o.value === run.settings.effort,
      ) &&
      (!run.settings.serviceTier ||
        selection?.options?.some(
          (o: any) =>
            o.id === "serviceTier" && o.value === run.settings.serviceTier,
        )) &&
      state.thread?.runtime_mode === run.settings.runtimeMode &&
      state.thread?.interaction_mode === run.settings.interactionMode &&
      (state.thread?.worktree_path || root) === run.directory;
    const live =
      !state.thread?.deleted_at &&
      state.turns.some((t: any) => t.started_at) &&
      !state.session?.last_error;
    if (!matches || !live)
      throw new Error(
        `Thread not verified: ${JSON.stringify({ thread: state.thread, session: state.session, turns: state.turns })}`,
      );
    run.status =
      run.status === "captured" ? "captured" : "awaiting-human-conversation";
    await saveJSON(join(dir, "run.json"), run);
    console.log(
      JSON.stringify(
        {
          threadId: run.threadId,
          title: state.thread.title,
          selection,
          session: state.session,
          messages: state.messages.length,
          firstAssistant:
            state.messages.find((m) => m.role === "assistant")?.text ?? null,
        },
        null,
        2,
      ),
    );
  } else if (command === "capture") {
    // The coordinator invokes this only after the human returns and says they are done.
    if (!run.threadId) throw new Error("No thread to capture");
    const connection = db();
    let evidence: any;
    try {
      evidence = connection.transaction(() => ({
        thread: connection
          .query(
            "SELECT thread_id, title, model_selection_json, runtime_mode, interaction_mode FROM projection_threads WHERE thread_id = ?",
          )
          .get(run.threadId),
        session: connection
          .query(
            "SELECT status, active_turn_id, last_error FROM projection_thread_sessions WHERE thread_id = ?",
          )
          .get(run.threadId),
        messages: connection
          .query(
            "SELECT message_id, turn_id, role, text, is_streaming, created_at, updated_at, attachments_json FROM projection_thread_messages WHERE thread_id = ? ORDER BY created_at, rowid",
          )
          .all(run.threadId),
        turns: connection
          .query(
            "SELECT turn_id, state, requested_at, started_at, completed_at FROM projection_turns WHERE thread_id = ? ORDER BY requested_at, row_id",
          )
          .all(run.threadId),
      }))();
    } finally {
      connection.close();
    }
    if (!evidence.thread) throw new Error("Thread not found");
    const pendingTurns = evidence.turns.filter(
      (t: any) => !t.completed_at && !t.started_at,
    );
    const acceptIdlePending =
      process.argv.includes("--include-pending") &&
      evidence.session?.status === "ready";
    if (
      evidence.session?.active_turn_id ||
      evidence.messages.some((m: any) => m.is_streaming) ||
      evidence.turns.some(
        (t: any) => !t.completed_at && (t.started_at || !acceptIdlePending),
      )
    )
      throw new Error(
        "Conversation still has an unfinished turn; wait for it to finish before capturing",
      );
    if (evidence.messages.filter((m: any) => m.role === "user").length < 2)
      throw new Error(
        "Only the runner's opening prompt exists; wait for the human conversation",
      );
    const captures = (await readdir(dir)).filter((f) =>
      /^capture-\d+$/.test(f),
    );
    const capture = `capture-${Math.max(0, ...captures.map((f) => Number(f.split("-")[1]))) + 1}`;
    const target = join(dir, capture);
    await mkdir(target);
    const raw = JSON.stringify(evidence, null, 2) + "\n";
    await writeFile(join(target, "transcript.json"), raw);
    const messages = evidence.messages.map(
      (m: any, i: number) =>
        `## ${i + 1}. ${m.role}${i === 0 ? " (runner kickoff)" : ""}\n\nMessage: ${m.message_id} · ${m.created_at}\n\n${m.text}\n`,
    );
    await writeFile(
      join(target, "transcript.md"),
      `# ${run.name}\n\nThread: ${run.threadId}\n\n${messages.join("\n")}`,
    );
    await saveJSON(join(target, "evidence.json"), {
      capturedAt: new Date().toISOString(),
      threadId: run.threadId,
      sha256: hash(raw),
      source: "T3 projection_thread_messages; turn and session metadata",
      coverage:
        "Stored chat messages including kickoff; tool activity and hidden reasoning are not exported. Attachments remain source references.",
      messages: evidence.messages.length,
      pendingTurns: pendingTurns.length,
      captureNote: pendingTurns.length
        ? "Human accepted the visible conversation as-is. The idle session retains unstarted pending turn metadata; preserve it without assuming those turns completed."
        : null,
    });
    run.status = "captured";
    run.latestCapture = capture;
    await saveJSON(join(dir, "run.json"), run);
    console.log(
      `Captured ${target}. Discuss the human's experience before proposing skill edits.`,
    );
  } else throw new Error(`Unknown command: ${command}`);
}

File: video-edit-workspace/eval.py
#!/usr/bin/env python3
"""Coordinate an ordinary video-production thread; keep evidence outside its project."""
import argparse
import datetime
import hashlib
import json
from pathlib import Path
import sqlite3
import subprocess
import sys

ROOT = Path(__file__).resolve().parent.parent
BASE = ROOT / 'video-edit-workspace/evals'


def save(path, value):
    path.write_text(json.dumps(value, indent=2, ensure_ascii=False) + '\n')


def sha(path):
    h = hashlib.sha256()
    with path.open('rb') as f:
        for b in iter(lambda: f.read(8 * 1024 * 1024), b''):
            h.update(b)
    return h.hexdigest()


def now():
    return datetime.datetime.now(datetime.timezone.utc).isoformat()


def inspect(thread):
    database = Path.home() / '.t3/userdata/state.sqlite'
    con = sqlite3.connect('file:' + str(database) + '?mode=ro', uri=True)
    con.row_factory = sqlite3.Row
    try:
        con.execute('BEGIN')
        one = lambda sql: next((dict(r) for r in con.execute(sql, (thread,))), None)
        many = lambda sql: [dict(r) for r in con.execute(sql, (thread,))]
        return {
            'thread': one('SELECT thread_id,title,worktree_path,model_selection_json,runtime_mode,interaction_mode,deleted_at FROM projection_threads WHERE thread_id=?'),
            'session': one('SELECT status,provider_name,provider_instance_id,provider_thread_id,active_turn_id,last_error FROM projection_thread_sessions WHERE thread_id=?'),
            'turns': many('SELECT turn_id,state,started_at,completed_at FROM projection_turns WHERE thread_id=? ORDER BY requested_at,row_id'),
            'messages': many('SELECT message_id,turn_id,role,text,is_streaming,created_at,updated_at,attachments_json FROM projection_thread_messages WHERE thread_id=? ORDER BY created_at,rowid'),
        }
    finally:
        con.rollback()
        con.close()


def verify(run):
    project = Path(run['directory'])
    for relative, expected in run['package_hashes'].items():
        if sha(project / relative) != expected:
            raise RuntimeError('Participant package changed: ' + relative)


def main():
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument('command', choices=['prepare','start','status','capture'])
    p.add_argument('iteration')
    p.add_argument('--directory', type=Path)
    p.add_argument('--prompt-file', type=Path)
    p.add_argument('--thread-helper', type=Path)
    p.add_argument('--provider', default='codex')
    a = p.parse_args()
    if not a.iteration.startswith('iteration-') or not a.iteration[10:].isdigit():
        p.error('Use iteration-N')
    out = BASE / a.iteration
    if a.command == 'prepare':
        if not all([a.directory,a.prompt_file,a.thread_helper]):
            p.error('prepare needs directory, prompt-file, and thread-helper')
        project = a.directory.resolve()
        prompt = a.prompt_file.read_text()
        # Test identities and rubric stay in this coordinator directory.
        if any(x in prompt.lower() for x in ['eval','benchmark','under test']):
            raise RuntimeError('Use an ordinary production prompt')
        out.mkdir(parents=True, exist_ok=False)
        (out/'prompt.txt').write_text(prompt)
        initial = out/'project-inputs'
        initial.mkdir()
        for name in ['AGENTS.md','DESIGN.md','TOOLS.md','.codex/config.toml','package.json','bun.lock']:
            src=project/name
            if src.exists():
                dest=initial/name;dest.parent.mkdir(parents=True,exist_ok=True);dest.write_bytes(src.read_bytes())
        package_hashes = {str(f.relative_to(project)):sha(f) for f in sorted((project/'.agents/skills').rglob('*')) if f.is_file()}
        run = {'created_at':now(),'name':'OBS September 19 · Educational edit','status':'prepared','directory':str(project),'branch':'main',
               'source_revision':subprocess.check_output(['git','rev-parse','HEAD'],cwd=ROOT,text=True).strip(),
               'project_initial_revision':subprocess.check_output(['git','rev-parse','HEAD'],cwd=project,text=True).strip(),
               'settings':{'provider':a.provider,'model':'gpt-6-astra','effort':'high','service_tier':'default','runtime_mode':'full-access','harness':'T3 Code'},
               'thread_helper':str(a.thread_helper.resolve()),'thread_helper_sha256':sha(a.thread_helper),'prompt_sha256':sha(out/'prompt.txt'),
               'package_hashes':package_hashes,'inputs':[{'name':f.name,'bytes':f.stat().st_size,'sha256':sha(f)} for f in sorted((project/'raw').iterdir()) if f.is_file()],
               'thread_id':None,'format':'landscape','format_basis':'Coordinator stated a landscape working assumption after optional format question; no human format selection received before preparation.',
               'human_review':'pending','isolation':'Standalone project; task-only prompt; six copied production skills; workspace-scoped disabling of ambient skills/plugins. Full-access remains enabled; this is context isolation, not a filesystem sandbox.'}
        save(out/'run.json',run)
        print(json.dumps({'status':run['status'],'directory':str(project),'inputs':[i['name'] for i in run['inputs']]}))
        return
    run=json.loads((out/'run.json').read_text())
    if a.command == 'start':
        if run['status']!='prepared' or run['thread_id']:
            raise RuntimeError('Launch already attempted. Inspect status before any retry.')
        verify(run)
        if sha(out/'prompt.txt')!=run['prompt_sha256'] or sha(Path(run['thread_helper']))!=run['thread_helper_sha256']:
            raise RuntimeError('Launch inputs changed')
        s=run['settings']
        command=[sys.executable,run['thread_helper'],'--name',run['name'],'--prompt',(out/'prompt.txt').read_text(),'--project-directory','/Users/asher/Projects/asher-skills','--directory',run['directory'],'--branch',run['branch'],'--provider',s['provider'],'--effort-option-id','reasoningEffort','--model',s['model'],'--effort',s['effort'],'--service-tier',s['service_tier'],'--runtime-mode',s['runtime_mode']]
        run['status']='launch_attempted';save(out/'run.json',run)
        result=subprocess.run(command,capture_output=True,text=True)
        (out/'dispatch.stdout').write_text(result.stdout);(out/'dispatch.stderr').write_text(result.stderr)
        try:
            dispatch=json.loads(result.stdout);run['thread_id']=dispatch.get('thread_id');save(out/'dispatch.json',dispatch)
        except ValueError:
            dispatch={}
        run['dispatch_exit_code']=result.returncode
        run['status']='start_acknowledged' if result.returncode==0 else 'launch_uncertain'
        save(out/'run.json',run)
        print(json.dumps({'status':run['status'],'thread_id':run['thread_id'],'detail':dispatch}))
        if result.returncode:raise SystemExit(result.returncode)
        return
    if not run['thread_id']:raise RuntimeError('No recorded thread identity')
    evidence=inspect(run['thread_id'])
    if a.command=='status':
        print(json.dumps({'thread':evidence['thread'],'session':evidence['session'],'turns':evidence['turns'],'recent_messages':evidence['messages'][-3:]},indent=2))
        return
    if (evidence['session'] or {}).get('active_turn_id') or any(m['is_streaming'] for m in evidence['messages']):
        raise RuntimeError('Participant still running; capture only after the turn settles')
    capture=out/('capture-'+str(len(list(out.glob('capture-*')))+1));capture.mkdir()
    save(capture/'thread.json',evidence)
    (capture/'transcript.md').write_text('\n\n'.join('## '+m['role']+' · '+m['message_id']+'\n\n'+m['text'] for m in evidence['messages']))
    project=Path(run['directory'])
    artifacts=[]
    for sub in ['output','work']:
        for f in sorted((project/sub).rglob('*')):
            if f.is_file():artifacts.append({'path':str(f.relative_to(project)),'bytes':f.stat().st_size,'sha256':sha(f)})
    save(capture/'artifacts.json',artifacts)
    verify(run)
    run['status']='captured';run['last_capture']=str(capture.relative_to(ROOT));run['human_review']='pending';save(out/'run.json',run)
    print(json.dumps({'capture':str(capture),'artifacts':len(artifacts),'human_review':'pending'}))


if __name__ == '__main__':main()
